src/garage/tf/algos/rl2.py

```diff
@@ -1,15 +1,340 @@
+                                 env_infos=env_infos,
+                                 paths=concatenated_path_in_meta_batch,
+                                 average_return=np.mean(undiscounted_returns))
+
+        return concatenated_path
+
+    def _concatenate_paths(self, paths):
+        """Concatenate paths.
+
+        The input paths are from different rollouts but same task/environment.
+        In RL^2, paths within each meta batch are all concatenate into a single
+        path and fed to the policy.
+
+        Args:
+            paths (dict): Input paths. All paths are from different rollouts,
+                but the same task/environment.
+
+        Returns:
+            dict: Concatenated paths from the same task/environment. Shape of
+                values: :math:`[max_path_length * episode_per_task, S^*]`
+            list[dict]: Original input paths. Length of the list is
+                :math:`episode_per_task` and each path in the list has
+                values of shape :math:`[max_path_length, S^*]`
+
+        """
+        returns = []
+
+        if self._flatten_input:
+            observations = np.concatenate([
+                self._env_spec.observation_space.flatten_n(
+                    path['observations']) for path in paths
+            ])
+        else:
+            observations = np.concatenate(
+                [path['observations'] for path in paths])
+        actions = np.concatenate([
+            self._env_spec.action_space.flatten_n(path['actions'])
+            for path in paths
+        ])
+        rewards = np.concatenate([path['rewards'] for path in paths])
+        dones = np.concatenate([path['dones'] for path in paths])
+        valids = np.concatenate(
+            [np.ones_like(path['rewards']) for path in paths])
+        returns = np.concatenate([path['returns'] for path in paths])
+
+        env_infos = np_tensor_utils.concat_tensor_dict_list(
+            [path['env_infos'] for path in paths])
+        agent_infos = np_tensor_utils.concat_tensor_dict_list(
+            [path['agent_infos'] for path in paths])
+        lengths = [path['lengths'] for path in paths]
+
+        concatenated_path = dict(
+            observations=observations,
+            actions=actions,
+            rewards=rewards,
+            dones=dones,
+            valids=valids,
+            lengths=lengths,
+            returns=returns,
+            agent_infos=agent_infos,
+            env_infos=env_infos,
+        )
+        return concatenated_path
+
+    @property
+    def policy(self):
+        """Policy.
+
+        Returns:
+            garage.Policy: Policy to be used.
+
+        """
+        return self._inner_algo.policy
+
+    @property
+    def max_path_length(self):
+        """Max path length.
+
+        Returns:
+            int: Maximum path length in a trajectory.
+
+        """
+        return self._max_path_length
+
+
+def stack_paths(max_len, paths):
```

Comment:
seems like this should go into `garage.tf.utils`?

Reply:
I think this function might already exist. Also, it should definitely be in `garage.misc.tensor_utils`, there's nothing TensorFlow specific here.

Reply:
fixed